tiledb/sm/query/sparse_index_reader_base.cc

```diff
@@ -181,67 +260,382 @@ Status SparseIndexReaderBase::load_initial_data() {
+  const auto dim_num = array_schema_->dim_num();
+  const auto cell_order = array_schema_->cell_order();
+
+  // No subarray set, return.
+  if (!subarray_.is_set()) {
+    return Status::Ok();
+  }
+
+  // Compute the number of non-default dimensions.
+  unsigned non_default_dims = 0;
+  for (unsigned d = 0; d < dim_num; d++) {
+    if (!subarray_.is_default(d))
+      non_default_dims++;
+  }
+
+  // If we have more than one non default dim, we'll need one temp bitmap per
+  // thread.
+  tdb_unique_ptr<ResourcePool<std::vector<uint64_t>>> bitmap_pool;
+  if (non_default_dims > 1) {
+    const auto num_threads =
+        storage_manager_->compute_tp()->concurrency_level();
+    bitmap_pool.reset(
+        tdb_new(ResourcePool<std::vector<uint64_t>>, num_threads));
+  }
+
+  // Process all tiles in parallel.
+  auto result_tiles_it = ResultTileIt(result_tiles);
+  auto status = parallel_for(
+      storage_manager_->compute_tp(), 0, result_tiles_it.size(), [&](uint64_t) {
+        // Take a result tile.
+        auto rt = result_tiles_it.get_next();
+
+        // Bitmap was already computed for this tile.
+        if (rt->bitmap_num_cells != std::numeric_limits<uint64_t>::max()) {
+          return Status::Ok();
+        }
+
+        // Make sure we have a bitmap available if there is more than one
+        // non default dimensions.
+        typedef ResourceGuard<std::vector<uint64_t>, ResourcePool>
+            Uint64VectorPool;
+        tdb_unique_ptr<Uint64VectorPool> guard = nullptr;
+        if (non_default_dims > 1)
+          guard.reset(tdb_new(Uint64VectorPool, *bitmap_pool.get()));
+
+        // Compute bitmaps one dimension at a time.
+        bool use_pool_buffer = false;
+        for (unsigned d = 0; d < dim_num; d++) {
+          bool tile_used = false;
+
+          // For col-major cell ordering, iterate the dimensions
+          // in reverse.
+          const unsigned dim_idx =
+              cell_order == Layout::COL_MAJOR ? dim_num - d - 1 : d;
+
+          // No need to compute bitmaps for default dimensions.
+          if (subarray_.is_default(dim_idx))
+            continue;
+
+          // Process all ranges at once.
+          // Note: this could be optimized. However, the result count is
+          // slowly going to get phased out so for now this code is going
+          // for simplicity.
+          rt->bitmap.resize(rt->cell_num());
+
+          // Make sure the pool buffer is ready for use.
+          if (use_pool_buffer) {
+            auto& pool_buffer = guard->get();
+            uint64_t memset_length =
+                std::min((uint64_t)pool_buffer.size(), rt->cell_num());
+            memset(pool_buffer.data(), 0, memset_length * sizeof(uint64_t));
+            pool_buffer.resize(rt->cell_num());
+          }
+
+          // Process all ranges in parallel.
+          const auto& ranges_for_dim = subarray_.ranges_for_dim(dim_idx);
+          std::mutex range_mtx;
+          auto status = parallel_for(
```

Comment:
I think we will want to eventually push the ranges deeper into the filtering inside the tile. Now we are making too many passes over the tile for a large number of ranges.